src/tiledbsoma_ml/query_ids.py

```diff
@@ -0,0 +1,198 @@
+        self,
+        *fracs: float,
+        seed: int | None = None,
+        method: SamplingMethod = "stochastic_rounding",
+    ) -> Tuple["QueryIDs", ...]:
+        """Split this |QueryIDs| into 1 or more |QueryIDs|, randomly sampled according ``fracs``.
+
+        - ``fracs`` must sum to $1$
+        - ``seed`` is optional
+        - ``method``: see |SamplingMethod| for details
+        """
+        split_fracs = np.cumsum(fracs)
+        assert fracs and np.isclose(split_fracs[-1], 1.0), "Fractions must sum to 1"
+
+        obs_joinids = self.obs_joinids
+        n_obs = len(obs_joinids)
+
+        rng = np.random.default_rng(seed)
+        shuffled_joinids = rng.permutation(obs_joinids)
+
+        if method == "deterministic":
+            split_idxs = np.round(split_fracs * n_obs).astype(int)
+        elif method == "multinomial":
+            split_idxs = np.cumsum(rng.multinomial(n_obs, np.array(fracs)))
+        elif method == "stochastic_rounding":
+            fracs_arr = np.array(fracs) * n_obs
+            split_bases = fracs_arr.astype(int)
+            split_idxs = split_bases.cumsum()
+            remainder_fracs = fracs_arr - split_bases
+            n = len(fracs)
+            remainders = np.zeros(n, dtype=int)
+            total_remainders = round(np.sum(remainder_fracs))
+            while total_remainders > 0:
+                pvals = remainder_fracs / np.sum(remainder_fracs)
+                choice = rng.choice(n, p=pvals)
+                remainders[choice] += 1
+                remainder_fracs[choice] = 0
+                total_remainders -= 1
+            split_idxs += remainders.cumsum()
+        else:
+            raise ValueError(f"Unknown sampling method: {method}")
+
+        splits = [
+            np.sort(split)
+            for split in np.array_split(shuffled_joinids, split_idxs[:-1])
+        ]
+        return tuple(evolve(self, obs_joinids=split) for split in splits)
+
+    def partitioned(
+        self,
+        partition: Optional[Partition] = None,
+    ) -> "QueryIDs":
+        """Create a new |QueryIDs| with |Q.obs_joinids| corresponding to a given GPU/worker |Partition|.
+
+        If ``None`` is provided, world size, rank, num workers, and worker ID will be inferred using helper functions
+        that read env vars (see |get_distributed_world_rank|, |get_worker_world_rank|).
+
+        When ``WORLD_SIZE > 1``, each GPU will receive the same number of samples (meaning up to ``WORLD_SIZE-1``
+        samples may be dropped).
+        """
+        if self.partition:
+            raise ValueError(
+                f"QueryIDs has already been partitioned ({self.partition})"
+            )
+
+        obs_joinids = self.obs_joinids
+        if partition:
+            rank = partition.rank
+            world_size = partition.world_size
+            worker_id = partition.worker_id
+            n_workers = partition.n_workers
+        else:
+            rank, world_size = get_distributed_world_rank()
```

Comment:
this code path appears to be unused.  Recommend removing it (and removing the optionality on the partition param).  Or add a test case for it :-)